### PR TITLE
Kayla Ecker - Carets

### PR DESCRIPTION
**Pull request by kaylaecker**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | *PROJECT still a work in progress, but wanted to submit a timely PR. will update comment when finished* Most of my methods were not in the model this time. 

Describe how you approached testing that model method. What edge cases did you come up with? | Because most of my methods are in the controller, most of my testing is too. However, under model 'work' I do have a must have a title test, and checked to see that errors keys included title and category.  

Describe an edge case test you wrote for a controller | Under controller 'work' I checked for things such as items must respond with success or failure depending on the outcome. Also checked that publication years/descriptions were equal to what they were set as, and things of that nature.|

What are `session` and `flash`? What is the difference between them? | Flash will only last one more request, while session will last as long as the cookie does not expire. Session is a hash-like object that allows you to maintain requests without necessarily using the database, useful for keeping the id of a user within the session. Flash is similar, but the data persists for only one request (i.e., for a display that something was successful) |

Describe a controller filter you wrote. |
What was one thing that you gained more clarity on through this assignment? | Session and flash, and the general set up. I had major issues with attempting to use branches, and ended up accidentally deleting my work instead of merging. 

What is the Heroku URL of your deployed application | Still working on the css before deployment|

Do you have any recommendations on how we could improve this project for the next cohort? | This project should have been the pair project, not last weeks (which was MUCH quicker, and could easily have been done alone. This was more involved.). There were not enough days to get it done, especially since this was a shorter week than normal. Expecting that we will be able to do the whole website including the designing per the specs within the time frame is not reasonable, particularly since many of us do have outside responsibilities. If the learning goals were to do testing and such, that should have been the focus not all of these other things.  |

### CODE REVIEW

**Comment by tildeee** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | x
Agile practices (trello board) | no trello board attached
Comprehension questions	| x
**General** |
Rails fundamentals (RESTful routing, use of named paths) | could have cleaned up your routes.rb file
Semantic HTML | See attached comment about use of `<body>` tag
Errors are reported to the user | x
Business logic lives in the models | x
Models are thoroughly tested, including relations, validations and any custom logic | only on works
Controllers are thoroughly tested | only positive cases
**Wave 1 - Media** |
Splash page shows the three media categories | x
Basic CRUD operations on media are present and functional | x
**Wave 2 - Users and Votes** |
Users can log in and log out | see attached comment about logging in
The ID of the current user is stored in the session |
Individual user pages and the user list are present | no individual user pages through "view all users" page, no view for going to users through "works" show page
A user cannot vote for the same media more than once | x
All media lists are ordered by vote count | not ordered
Splash page contains a media spotlight | logic is `(Work.all).first`, not by vote
Media pages contain lists of voting users | x
**Wave 3 - Styling** |
Foundation is used appropriately | did not load Foundation
Look and feel is similar to the original | similar header
**Overall** |

**Comment by tildeee** [comment]
- controller tests fail
- when upvoting, votes create method doesn't redirect
- log in doesn't work-- there's a typo, I'm adding a comment on this
- log out isn't implemented

File: app/views/sessions/log_form.html.erb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+<h2>Log In</h2>
+
+<%= form_tag do %>
+<%= label_tag(:username, "Username") %>
+<%= text_field_tag(:username) %>
```
Comment (tildeee) [inline app/views/sessions/log_form.html.erb:5]: logging in doesn't work as is-- if you change this form to use the symbol `:name` instead of `:username` it does!

The symbol you specify here (in this case, `:username`) gets stored into the params hash. This params hash is what is used while creating a user (see `sessions_controller.rb` in line 10). I'm adding a comment at that point too

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
+class SessionsController < ApplicationController
+  def login
+    @user = User.find_by(name: params[:name])
+    if @user
+      flash[:success] = "Logged in as #{@user.name}"
+      session[:user_id] = @user.id
+      session[:user] = @user.name
+      redirect_to root_path
+    else
+      @user = User.new(name: params[:name])
```
Comment (tildeee) [inline app/controllers/sessions_controller.rb:10]: this line says "take whatever is stored in `params[:name]` and then put it into the key `name` for a new `User`.

In the log_form, params gets the value from the text field and places it into `params[:username]`, so at this point in the controller, `params[:name]` is nil and the user doesn't save.

File: app/views/works/_form.html.erb [hunk 3]
```diff
@@ -0,0 +1,20 @@
+<body>
```
Comment (tildeee) [inline app/views/works/_form.html.erb:1]: This isn't how to use the `<body>` tag. The body tag is meant to contain all of the content HTML (in direct opposition to the `<head>` tag, which contains all the meta data about the site). A `<body>` tag only occurs once in a given HTML document.

Your application.html.erb layout already has a `<body>` tag, so you don't need to add it here.

File: app/views/works/_works.html.erb [hunk 4]
```diff
@@ -0,0 +1,28 @@
+</ul>
+<% works.each do |work| %>
+<body>
```
Comment (tildeee) [inline app/views/works/_works.html.erb:3]: see my other comment about the use of `<body>`
